firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java

```diff
@@ -0,0 +1,347 @@
+    }
+    CompositeFilter other = (CompositeFilter) o;
+    // Note: This comparison requires order of filters in the list to be the same, and it does not
+    // remove duplicate subfilters from each composite filter. It is therefore way less expensive.
+    return operator == other.operator && filters.equals(other.filters);
+  }
+
+  @Override
+  public int hashCode() {
+    int result = 37;
+    result = 31 * result + operator.hashCode();
+    for (Filter filter : filters) {
+      result += filter.hashCode();
+    }
+    return result;
+  }
+
+  /**
+   * Returns a new composite filter that contains all filter from `this` plus all the given filters
+   */
+  public CompositeFilter addFilters(List<Filter> otherFilters) {
+    List<Filter> mergedFilters = new ArrayList<>(filters);
+    mergedFilters.addAll(otherFilters);
+    return new CompositeFilter(mergedFilters, operator);
+  }
+
+  /**
+   * Returns a new composite filter that contains all filters from `this` plus the given field
+   * filter
+   */
+  public CompositeFilter addFilter(FieldFilter other) {
+    List<Filter> mergedFilters = new ArrayList<>(filters);
+    mergedFilters.add(other);
+    return new CompositeFilter(mergedFilters, operator);
+  }
+
+  /**
+   * Applies the associativity property to this filter to get a new filter and returns it.
+   *
+   * <p>For Example, it turns this: OR(A, OR(B, OR(C, D, OR(E, F), AND(G, OR(H, I))))
+   *
+   * <p>Into this: OR(A, B, C, D, E, F, AND(G, OR(H, I))
+   */
+  @NonNull
+  @Override
+  public Filter applyAssociativity() {
+    // Associativity for AND(X) should result in X.
+    if (filters.size() == 1) {
+      return filters.get(0).applyAssociativity();
+    }
+
+    // Associativity applied to a flat composite filter results in itself.
+    if (!containsCompositeFilters()) {
+      return this;
+    }
+
+    // First apply associativity to all subfilters. This will in turn recursively apply
+    // associativity to all nested composite filters and field filters.
+    List<Filter> updatedFilters = new ArrayList<>();
+    for (Filter subfilter : filters) {
+      updatedFilters.add(subfilter.applyAssociativity());
+    }
+
+    // Take out components of composite subfilters where possible.
+    List<Filter> newSubfilters = new ArrayList<>();
+    for (Filter subfilter : updatedFilters) {
+      if (subfilter instanceof FieldFilter) {
+        newSubfilters.add(subfilter);
+      } else if (subfilter instanceof CompositeFilter) {
+        CompositeFilter compositeSubfilter = (CompositeFilter) subfilter;
+        if (compositeSubfilter.getOperator() == operator) {
+          // Filter: (A | (B | C))
+          // compositeSubfilter: (B | C)
+          // Result: (A | B | C)
+          newSubfilters.addAll(compositeSubfilter.getFilters());
+        } else {
+          // Filter: (A | (B & C))
+          // compositeSubfilter: (B & C)
+          // Result: (A | (B & C))
+          newSubfilters.add(subfilter);
+        }
+      }
```

Comment:
Would removing this code change anything about the final DNF?

Reply:
yeah, it's not possible to reach DNF form if associativity is not applied at every iteration. Without this code the final form may not be DNF and if it is, it may contain nested OR conditions which we don't want.